### DISCUSSION TITLE
Would it help to not delete the struct entity on cancel or when the stream is fully withdrawn?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#258] [Ideas]
This is an open-ended question addressed to @razgraf and @gavriliumircea - would it help you guys if we did not delete the struct entity on cancel or when the stream is fully withdrawn?

- https://github.com/sablierhq/v2-core/blob/d9fffa978da342785be2bd3b5899a803e6b0a212/src/SablierV2Linear.sol#L288
- https://github.com/sablierhq/v2-core/blob/d9fffa978da342785be2bd3b5899a803e6b0a212/src/SablierV2Linear.sol#L432

The way I see it, this would be useful for two purposes:

1. Better data support in the light app.
2. Better data support for the web3 community who might prefer to query the data from the smart contracts directly, rather than the subgraphs.

But please do let me know if you envisage more use cases!

In terms of implementation, this would't be easy, since we would have to refactor the `streamExists` modifier into a `streamExistsAndNotEnded` modifier, and then add a slew of new tests that ensure that ended streams cannot be interacted with.

There would also be two additional implications:

1. The gas cost for canceling and withdrawing would increase a bit, because there would be no storage erase refund anymore.
2. Long term, we might incur additional costs on smart contract storage rent. It is known that ETH2 is going to update the current model, because it is unsustainable. It is not feasible to keep increasing the storage of the blockchain forever.

**Comment by razgraf** [comment]
The historical support for querying streams is the stronger argument, IMO, in this case. While with the subgraph (or any other caching mechanism) data lives on forever, I think for direct web3 calls, preserving that struct is quite valuable.

Firstly, it would help our light app a lot - anyone sending a stream id/configuration would get back results (about who sent the stream, how much was in that stream, maybe peek under the hood to see it's segments etc.). Secondly, it would probably make sense for out integrators too. Logically removing something (through something like the `end time < now` invariant to mark an ended stream) seems a bit more elegant than physically removing it from storage. 

I'd vote to keep it **IF** and only if the cost downside isn't too big. Regarding ETH2's storage updates, I think we're safe to say it won't happen in the next two/three years. Gives us plenty of time to understand it and react accordingly.

**Reply by PaulRBerg** [reply]
Thanks for sharing your thoughts. I must say, I am so much on the fence about this issue, because I agree with everything you said.

However, I've also thought about other implications in the meantime, and this refactor (if we end up doing it) won't be easy.

For instance, what would it then even mean to "cancel" a stream? Should a canceled stream be associated with a special property `isCanceled` that belongs to the struct? If not, and it should be associated with the same `isEnded` property that gets set to `true` when the stream is fully withdrawn, isn't this self-defeating, because there would be no way to tell if the stream had been canceled or fully withdrawn from without the subgraph?

Let me sleep more on this before making a decision.

> through something like the `end time < now` invariant to mark an ended stream

This is not what I was suggesting. Here's a more concrete description.

- Stream gets fully withdrawn -> `stream.isEnded` is set to `true`
- Stream gets canceled -> `stream.isEnded` is set to `true` (subject to change as explained above)

We would never change the state of the stream based on the current time.

**Reply by razgraf** [reply]
> We would never change the state of the stream based on the current time.

In my mind that was more of a view method, for the time invariant, not a state field. At any rate, I don't think we need a flag for an "ended" state, we can just get that manually, through easy calls for the `withdrawnAmount` and/or the `stopTime`.

For the cancel state, I think a flag would be great though.

Overall I do think this would improve the DX of our contracts in the long run. I get that the refactor may be problematic but I feel it helps with the logic and intuitiveness of our state (esp. in a historical context).

**Reply by PaulRBerg** [reply]
Well, if if we can easily obtain the time status through a call to `stopTime` and/ or `withdrawnAmount`, then I don't think we even need to bother about implementing a bespoke view method.

Re flag for cancel - yeah, there is no way to tell otherwise.

> I feel it helps with the logic and intuitiveness of our state (esp. in a historical context).

Agree. I will figure out something.

**Comment by PaulRBerg** [comment]
@razgraf You will like it to hear that I have managed to come up with a simple and elegant solution for avoiding the stream deletion. I have opened an issue to track its implementation (which I started working on today!):

https://github.com/sablierhq/v2-core/issues/269

Requests for feedback:

1. Are you cool with naming the nonexistent status of the stream `NULL`? Other ideas were `NONEXISTENT` and `UNCREATED`, though I liked `NULL` better. "Uncreated" has religious connotations.
2. Same question for `FINISHED`. The trouble here is that we want to refer to the fully withdrawn status of a stream, i.e. that state the stream is in once the recipient withdraws the entire `depositAmount`. I am not super happy with `FINISHED`, because it can be misinterpreted as meaning that the current time is greater than the stop time. Yet, I also can't think of any better alternative. Maybe `CLOSED`, or `SETTLED`? What do you think?

**Reply by razgraf** [reply]
My only "beef" with `NULL` is that it comes very close to the usual `null` reserved word. Although solidity doesn't have it (yet) and a full uppercase `NULL` or `UNDEFINED` isn't read as a lowercase reserved word by languages such as JS.. it's still weird.

On the other hand, it describes the functionality perfectly.
Other suggestions: `NIL`, `INVALID`, `EMPTY` (although this could be confused with a stream with no-funds left a.k.a. `FINISHED`), `UNDEFINED` if we have no issue with these words.

---

As far as `FINISHED` goes, maybe pick something with a numeric connotation? At the end of the day it should define the state of a stream with no funds left inside. 

Other suggestions: `DEPLETED`,`DRY`.

**Reply by PaulRBerg** [reply]
Enum values are actually encoded as `uint8` in Solidity, so you will not have to use "NULL" as a key in the client app. Thus, I don't think the potential clash with the reserved keyword is worrysome, since you can use tweak the enum name locally (in the worst case scenario).

Regarding a clash with Solidity - not a problem. Solidity will likely never have a `null` keyword. It will introduce a whole cascade of social coordination issues.

Re `NIL` - it suffers from the same potential keyword clash in other languages such as Swift. However, it is less clearer than `NULL`. According to Google Ngram, "null" is much more popular a word than "nil" is.

Regarding `DEPLETED` - I like it a lot. Certainly much clearer than `FINISHED`, which is guaranteed to be misunderstood for meaning that the stream ended chronologically.

<img width="1677" alt="Screenshot 2023-01-15 at 7 22 54 PM" src="https://user-images.githubusercontent.com/8782666/212556694-8aee7730-cdfc-4602-935c-dbc610dd8441.png">

**Comment by PaulRBerg** [comment]
It looks like Aave V3 has implemented a similar enum called [`InterestRateMode`](https://github.com/aave/aave-v3-core/blob/b007d1ee29ad2e6d9f2eb7459acb98a40855dbd5/contracts/protocol/libraries/types/DataTypes.sol#L82), and for the default value they have gone with `NONE`:

```solidity
enum InterestRateMode {
    NONE,
    STABLE,
    VARIABLE
}
```

WDYT, @razgraf? Do you like this better than `NULL`?

I may have just gotten used to `NULL`, but I like it better than `NONE`. `NONE` makes me think of something that *might* exist but it's just not present at the moment (as in "none found here, though things typically exist here").

**Reply by razgraf** [reply]
No strong opinion, pick whichever you like. "NONE" is less tech-ish, sounds a bit more natural than `NULL`, both do the job.

<img width="1511" alt="Screenshot 2023-01-28 at 13 49 06" src="https://user-images.githubusercontent.com/15740806/215264873-a0399545-3034-4db3-8260-052cf72bd47b.png">

**Reply by PaulRBerg** [reply]
Well, the audience in this context are devs - no end user shall ever know about the `NULL`/ `NONE` status.

I will stick with `NULL`.

**Reply by razgraf** [reply]
> Well, the audience in this context are devs

Isn't this conflicting with: https://github.com/sablierhq/v2-core/discussions/287#discussioncomment-4770833. When functions are seen by users, their results and the status will be shown as well.

**Reply by PaulRBerg** [reply]
Good catch of inconsistent stances.

I take back my statement about no end user ever seeing the status - some end users will eventually notice it, but `NULL` is fine. We will explain it in the documentation website.

**Comment by PaulRBerg** [comment]
Locking as this was implemented in #270 and #285 and we have settled the API.
